docs/concepts/clustering.rst

```diff
@@ -138,48 +137,50 @@ the next update.
 Master Node Election
 --------------------
 
-The process of electing a node as the master node in a cluster is called Master
-Node Election. There must be only one master node per cluster at any single
-time. In a CrateDB cluster any node is eligible to be elected as a master node,
-although this could also be restricted to a subset of nodes if required. These
-nodes will then elect a single node as master to coordinate the cluster state
-across the cluster.
-
+In a CrateDB cluster there can only be one master node at any single time.
+The cluster only becomes available to serve requests once a master has been 
+elected, and a new election takes place if the current master node becomes 
+unavailable.
+By default all nodes are master-eligible, but 
+:ref:`a node setting <node.master>`
+is available to indicate, if desired, that a node must not take on the role
+of master. To elect a master among the eligible nodes, a majority 
+(``floor(half)+1``), also known as *quorum*, is required among a subset of 
+all master-eligible nodes, this subset of nodes is known as the
+*voting configuration*.
+The *voting configuration* is a list which is persisted as part of the cluster
+state. It is maintained automatically in a way that makes so that split-brain
+scenarios are never possible.
+Every time a node joins the cluster, or leaves the cluster, even if it is 
+for a few seconds, CrateDB re-evaluates the voting configuration.
+If the new number of master-eligible nodes in the cluster is odd, CrateDB will
+put them all in the voting configuration.
+If the number is even, CrateDB will exclude one of the master-eligible nodes
+from the voting configuration.
+A master node rescinds its role if it cannot contact a quorum of nodes from
+the latest voting configuration.
+
+.. WARNING::
+
+   If you do infrastructure maintenance, please note that as nodes are shutdown 
+   or rebooted, they will temporarily leave the voting configuration, and for
+   the cluster to elect a master a quorum is required among the
+   nodes that were last in the voting configuration. For instance, if you
+   have a 5-nodes cluster, with all nodes master-eligible, and node 1 is 
+   currently the master, and you shutdown node 5, then node 4, then node 3, 
+   the cluster will stay available as the voting configuration will have 
+   adapted to only have nodes 1 and 2 on it.
+   If you then shutdown one more node the cluster will become unavailable as
+   a quorum of 2 out of nodes 1 and 2 is now required and not available.
+   To bring the cluster back online at this point you will require both nodes 1
```

Comment:
Is this correct? since all nodes are master eligible, why do we need both nodes 1 & 2 to become available?

Reply:
I am a bit confused, could you be more clear what you tested and how? logs would also be interesting.

Reply:
Hi, I did further tests and amended my last comment.

My latest test was like this, thank you for reviewing:
5 nodes (individual VMs - not containers):

|   hostname     |            id           |
|----------------|-------------------------|
| ubuntuserver1  |  Hj0j4oBYRz-X7OVqs6Q9fA |
| ubuntuserver2  |  YtWE40eKRHavYefEYLdaKA |
| ubuntuserver3  |  xOJlD0ayT2a6fskzlJ072A |
| ubuntuserver4  |  GTUJcoWjQ7m3-4J4Pk66ag |
| ubuntuserver5  |  Fhz9cxNeTDuFhqAslQctjQ |

ubuntuserver4 is the master

`systemctl stop crate` on 1, then on 2, then on 3, then on 4, then on 5
waited a bit

Then started one node at a time and let it run until it logged the "an election requires at least" message, took note, shut it down, and moved to the next node:

```
[2023-03-06T15:18:12,740][WARN ][o.e.c.c.ClusterFormationFailureHelper] [ubuntuserver1] master not discovered or elected yet, an election requires at least 3 nodes with ids from [YtWE40eKRHavYefEYLdaKA, Fhz9cxNeTDuFhqAslQctjQ, xOJlD0ayT2a6fskzlJ072A, Hj0j4oBYRz-X7OVqs6Q9fA, GTUJcoWjQ7m3-4J4Pk66ag]
```
--> this makes sense, when 1 was shutdown we had a 5 nodes cluster, 3 were needed for quorum

```
[2023-03-06T15:19:11,832][WARN ][o.e.c.c.ClusterFormationFailureHelper] [ubuntuserver2] master not discovered or elected yet, an election requires at least 2 nodes with ids from [Fhz9cxNeTDuFhqAslQctjQ, YtWE40eKRHavYefEYLdaKA, GTUJcoWjQ7m3-4J4Pk66ag]
```
--> this makes sense, when 2 was shutdown we had a 4 nodes cluster (2345), 3 appears to have been excluded from the voting configuration.

```
[2023-03-06T15:13:38,309][WARN ][o.e.c.c.ClusterFormationFailureHelper] [ubuntuserver3] master not discovered or elected yet, an election requires at least 2 nodes with ids from [Fhz9cxNeTDuFhqAslQctjQ, xOJlD0ayT2a6fskzlJ072A, GTUJcoWjQ7m3-4J4Pk66ag]
```
--> this makes sense, when 3 was shutdown we had a 3 nodes cluster (345)

```
[2023-03-06T15:19:46,906][WARN ][o.e.c.c.ClusterFormationFailureHelper] [ubuntuserver4] master not discovered or elected yet, an election requires at least 2 nodes with ids from [Fhz9cxNeTDuFhqAslQctjQ, xOJlD0ayT2a6fskzlJ072A, GTUJcoWjQ7m3-4J4Pk66ag]
```

```
[2023-03-06T15:20:18,012][WARN ][o.e.c.c.ClusterFormationFailureHelper] [ubuntuserver5] master not discovered or elected yet, an election requires at least 2 nodes with ids from [Fhz9cxNeTDuFhqAslQctjQ, xOJlD0ayT2a6fskzlJ072A, GTUJcoWjQ7m3-4J4Pk66ag
```
--> looking at this, and at other tests I did, it would seem that **perhaps when going from 3 nodes to 2 the voting configuration does not shrink anymore?**

Another observation is that these messages that read *"an election requires at least"* seems to be produced by each node individually looking at the last state of the voting configuration they knew, but an election only happens when all the nodes involved are satisfied, which is good. For instance node 1 would be satisfied with 123 and 4 and 5 would be satisfied with 4 and 5, but because 2 and 3 require either 4 or 5 there is no risk of split brain (tested again).

If you @proddata or @matriv could confirm the question in bold I will amend the paragraph accordingly.

Reply:
https://github.com/crate/crate/blob/c9de95d9741c0eb22c24c7f491548342be4efc24/server/src/main/java/org/elasticsearch/cluster/coordination/Reconfigurator.java#L52-L55

Reply:
> systemctl stop crate on 1, then on 2, then on 3, then on 4, then on 5
did you wait a after shutting them down or did you do it all at once?

Reply:
-- master state x
- node 1 was part of a cluster with 5 nodes -> voting configuration includes all 5 master eligible nodes  [1,2,3,4,5]
- node 1 gets removed -> cluster has 4 nodes
--> The voting configuration gets adjusted to include 3 nodes [2,4,5]
--> master state x+1
- node 2 gets removed -> cluster has 3 nodes
--> The voting configuration gets adjusted to include 3 nodes [3,4,5]
--> master state x+2

---

Removing node 3 should remove it from the voting configuration, but I assume this was done in a too quick manner

Reply:
That line limits only the size of the configuration, but doesn't tell much about the content.
Why a removed node counts as "live" is a bit questionable, but that is probably how it is.

To make it short: "To start a cluster you need a quorum of nodes for the voting configuration in the last committed cluster state."